### PR TITLE
Alternate fix for FITS regression with table keywords

### LINKED ISSUE #7145: New FITS table column attributes break existing code
Change in behaviour of binTableHDU.fromcolumns:

Version 2.0.3: passing a header containing keywords with names like "TCRPX2" to from_columns creates an hdu object with these keywords in the header

Version 3.dev: passing a header with such keywords results in them not making it to the header of the output

This code demonstrates:

```python
#! /usr/bin/env python
import astropy
from astropy.io import fits
import numpy as np

def test_fromcolumns():
    print(astropy.__version__)
    cd = create_column_descriptions()
    header = create_header()

    print("Header that is input to fits.BinTableHDU.from_columns:")
    print(header)

    hdu = fits.BinTableHDU.from_columns(cd, header)

    print("Header of hdu object returned from fits.BinTableHDU.from_columns:")
    print(hdu.header)

    return

def create_column_descriptions():
    col = []
    col.append(fits.Column(name="TIME", format="1E", unit="s"))
    col.append(fits.Column(name="RAWX", format="1I", unit="pixel"))
    col.append(fits.Column(name="RAWY", format="1I", unit="pixel"))
    cd = fits.ColDefs(col)

    return cd

def create_header():
    hdr = fits.Header()
    hdr['RA'] = (1.581250000000E+00, 'RA of reference aperture center ')
    hdr['DEC'] = (2.020277777778E+01, 'Declination of reference aperture center')
    hdr['TCTYP2'] = ('RA---TAN', 'axis type for dimension 1')
    hdr['TCTYP3'] = ('ANGLE   ', 'axis type for dimension 2')
    hdr['TCRVL2'] = (-999.0, 'sky coordinates of 1st axis')
    hdr['TCRVL3'] = (-999.0, 'sky coordinates of 2nd axis')
    hdr['TCRPX2'] = (1.0, 'axis1 pixel of tangent plane direction')
    hdr['TCRPX3'] = (1.0, 'axis2 pixel of tangent plane direction')
    hdr['TALEN2'] = (16384, 'length of axis 1')
    hdr['TALEN3'] = (1024, 'length of axis 2')
    hdr['TC2_2'] = (0.0, 'partial of first axis coordinate w.r.t. x')
    hdr['TC2_3'] = (0.0, 'partial of first axis coordinate w.r.t. y')
    hdr['TC3_2'] = (0.0, 'partial of second axis coordinate w.r.t. x')
    hdr['TC3_3'] = (0.0, 'partial of second axis coordinate w.r.t. y')
    hdr['TCUNI2'] = ('angstrom', 'units of first coordinate value')
    hdr['TCUNI3'] = ('deg     ', 'units of second coordinate value')

    return hdr

```

Edit: Added syntax highlighting.

### PR DESCRIPTION
**Pull request by astrofrog**
This is an alternative to https://github.com/astropy/astropy/pull/7153 to fix #7145 - basically it just prevents time-related keywords from being automatically removed from the Header. Time-related keywords can still be overwritten if the corresponding attributes are set on Column objects, but this doesn't affect backward-compatibility since those Column attributes didn't exist before. This approach thus guarantees full backward-compatibility (unless I've overlooked something), and emits a deprecation warning to mention that these keywords will be dropped in future. Example usage:

```python
In [1]: import astropy
   ...: from astropy.io import fits
   ...: import numpy as np
   ...: 
   ...: def create_column_descriptions():
   ...:     col = []
   ...:     col.append(fits.Column(name="TIME", format="1E", unit="s"))
   ...:     col.append(fits.Column(name="RAWX", format="1I", unit="pixel"))
   ...:     cd = fits.ColDefs(col)
   ...: 
   ...:     return cd
   ...: 
   ...: def create_header():
   ...:     hdr = fits.Header()
   ...:     hdr['RA'] = (1.581250000000E+00, 'RA of reference aperture center ')
   ...:     hdr['DEC'] = (2.020277777778E+01, 'Declination of reference aperture center')
   ...:     hdr['TCTYP2'] = ('RA---TAN', 'axis type for dimension 1')
   ...:     hdr['TCRVL2'] = (-999.0, 'sky coordinates of 1st axis')
   ...: 
   ...:     return hdr
   ...: 
   ...: columns = create_column_descriptions()
   ...: header = create_header()
   ...: 

In [2]: print(header.tostring(sep='\n'))
   ...: print(columns)
   ...: 
RA      =              1.58125 / RA of reference aperture center                
DEC     =       20.20277777778 / Declination of reference aperture center       
TCTYP2  = 'RA---TAN'           / axis type for dimension 1                      
TCRVL2  =               -999.0 / sky coordinates of 1st axis                    
END                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
ColDefs(
    name = 'TIME'; format = '1E'; unit = 's'
    name = 'RAWX'; format = '1I'; unit = 'pixel'
)
```

So far so good. Let's make a new HDU from the columns and header:

```python
In [3]: hdu = fits.BinTableHDU.from_columns(columns, header)
   ...: 
/Users/tom/Dropbox/Code/Astropy/astropy/astropy/io/fits/hdu/table.py:341: UserWarning: The following keywords are now recognized as special column-related attributes and should be set via the Column objects: TCRVLn, TCTYPn. In future, these values will be dropped from manually specified headers automatically and replaced with values generated based on the Column objects.
  "Column objects.".format(keys))
```

We see that this causes the above warning. The header still contains those keywords, and the columns don't (consistent with previous behavior):

```
In [4]: print(hdu.header.tostring(sep='\n'))
   ...: print(hdu.columns)
   ...: 
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                    6 / length of dimension 1                          
NAXIS2  =                    0 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    2 / number of table fields                         
RA      =              1.58125 / RA of reference aperture center                
DEC     =       20.20277777778 / Declination of reference aperture center       
TCTYP2  = 'RA---TAN'           / axis type for dimension 1                      
TCRVL2  =               -999.0 / sky coordinates of 1st axis                    
TTYPE1  = 'TIME    '                                                            
TFORM1  = '1E      '                                                            
TUNIT1  = 's       '                                                            
TTYPE2  = 'RAWX    '                                                            
TFORM2  = '1I      '                                                            
TUNIT2  = 'pixel   '                                                            
END                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
ColDefs(
    name = 'TIME'; format = '1E'; unit = 's'
    name = 'RAWX'; format = '1I'; unit = 'pixel'
)
```

Now let's write out the file and read in back in:

```python
In [5]: hdu.writeto('time.fits', overwrite=True)
   ...: 
   ...: hdu2 = fits.open('time.fits')[1]
   ...: 

In [6]: print(hdu2.header.tostring(sep='\n'))
   ...: print(hdu2.columns)
   ...: 
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                    6 / length of dimension 1                          
NAXIS2  =                    0 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    2 / number of table fields                         
RA      =              1.58125 / RA of reference aperture center                
DEC     =       20.20277777778 / Declination of reference aperture center       
TCTYP2  = 'RA---TAN'           / axis type for dimension 1                      
TCRVL2  =               -999.0 / sky coordinates of 1st axis                    
TTYPE1  = 'TIME    '                                                            
TFORM1  = '1E      '                                                            
TUNIT1  = 's       '                                                            
TTYPE2  = 'RAWX    '                                                            
TFORM2  = '1I      '                                                            
TUNIT2  = 'pixel   '                                                            
END                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
ColDefs(
    name = 'TIME'; format = '1E'; unit = 's'
    name = 'RAWX'; format = '1I'; unit = 'pixel'; coord_type = 'RA---TAN'; coord_ref_value = -999.0
)
```

At this point, the new column attributes are picked up (as expected) and they are still in the header as everything is now in sync. No warning was emitted when reading a file.

Let's just set a column attribute:

```python
In [9]: hdu2.columns[1].coord_type = 'DEC--TAN'

In [10]: hdu2.header['TCTYP2']
Out[10]: 'DEC--TAN'
```

The attributes in the header are now kept in sync and are not cleared by default when initializing from a header.

### CODE REVIEW

**Comment by mhvk** [comment]
I think this alternative is better, though do wonder about the specific warning. I think there may continue to be a case for adding specialized column keywords to headers for quite a while, so perhaps we should only warn if both the header and the column define a particular keyword? And the warning should be that such double definitions may lead to an exception in the future?

**Comment by pllim** [comment]
@stscirij , do you want to take this for a spin and see?

**Comment by stscirij** [comment]
Yes but I can't do it until tonight at the earliest


On Feb 3, 2018 10:41 AM, "P. L. Lim" <notifications@github.com> wrote:

@stscirij<https://github.com/stscirij> , do you want to take this for a spin and see?

**Comment by stscirij** [comment]
I ran my test with this version and it passed!!  Awesome!  Thank you for coming up with this fix!

**Comment by astrofrog** [comment]
@stscirij - thanks for confirming this!

What we need to decide is when to emit the warning and whether it should be re-worded (see @mhvk's comment above). Does anyone else have thoughts on this?

**Comment by astrofrog** [comment]
If we can't decide now, maybe it's safer to include the deprecation warning, as we can always decide to 'cancel' the deprecation later?

**Comment by saimn** [comment]
I think I prefer this approach since it does not try to read and set the attributes from the header. I just wonder if we can do everything if `_clear_table_keywords`. You current version has the benefit of raising the warning immediately, when calling `.from_columns`, but putting the verification in `_clear_table_keywords` would also allow to detect if the header has been modified after the HDU creation. Indeed, If I had to put keywords like this, I would probably first create the HDU, and then set the keywords.

**Comment by astrofrog** [comment]
@saimn - the issue is that in ``_clear_table_keywords`` it's not clear if keywords have been added by the user or automatically based on the column attributes.

**Comment by saimn** [comment]
@astrofrog - Ah yes, you're right. Looks good to me then, it just misses a test to be sure that the keywords are preserved ?

**Comment by astrofrog** [comment]
@saimn - ah yes, I'll add a test now!

File: astropy/io/fits/hdu/table.py [hunk 1]
```diff
@@ -243,7 +243,11 @@ class _TableBaseHDU(ExtensionHDU, _TableLikeHDU):
     data : array
         Data to be used.
     header : `Header` instance
-        Header to be used.
+        Header to be used. If the ``data`` is also specified, header keywords
+        specifically related to defining the table structure (such as the
+        "TXXXn" keywords like TTYPEn) will be overridden by the supplied column
+        definitions, but all other informational and data model-specific
+        keywords are kept.
```
Comment (astrofrog) [inline astropy/io/fits/hdu/table.py]: This was the case before, but wasn't documented, so just copied the relevant part from the docstring of ``from_columns``.

**Comment by astrofrog** [comment]
@saimn - does this look ok now?

**Comment by saimn** [comment]
@astrofrog - Yep looks good ! Thanks for tackling this, and merging since everybody seem happy with this solution.

**Comment by SaraOgaz** [comment]
Just for clarification, is the far future behavior that once these keywords are transferred to column attributes upon reading to the Table, that they won't be re-saved out into the header upon converting the table back to a FITS file?  I.e. in the far future if a FITS file header has these keywords they will not round-trip?

**Comment by saimn** [comment]
Once the table has been saved, these keywords will be read as proper Column attributes, and saved as well. Last part of @astrofrog 's example above shows this.

**Comment by taldcroft** [comment]
@astrofrog - thanks for jumping in and fixing this side effect of our GSoC project!

**Comment by SaraOgaz** [comment]
@saimn I can see that it's working now, but I was wondering about the "far future" when you guys don't have the weaning off solution that you backtracked to.  But I"m assuming from your answer that part of things will stay the same. :)

**Pull request merged**